File: src/main/java/info/skyblond/archivedag/service/impl/GroupService.java
package info.skyblond.archivedag.service.impl;

import info.skyblond.archivedag.model.DuplicatedEntityException;
import info.skyblond.archivedag.model.EntityNotFoundException;
import info.skyblond.archivedag.model.bo.GroupDetailModel;
import info.skyblond.archivedag.model.entity.GroupMetaEntity;
import info.skyblond.archivedag.model.entity.GroupUserEntity;
import info.skyblond.archivedag.repo.GroupMetaRepository;
import info.skyblond.archivedag.repo.GroupUserRepository;
import info.skyblond.archivedag.util.GeneralKt;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.LinkedList;
import java.util.List;

@Service
public class GroupService {
    private final PatternService patternService;
    private final GroupMetaRepository groupMetaRepository;
    private final GroupUserRepository groupUserRepository;

    public GroupService(PatternService patternService, GroupMetaRepository groupMetaRepository, GroupUserRepository groupUserRepository) {
        this.patternService = patternService;
        this.groupMetaRepository = groupMetaRepository;
        this.groupUserRepository = groupUserRepository;
    }

    @Transactional
    public void createGroup(String groupName, String owner) {
        if (!this.patternService.isValidGroupName(groupName)) {
            throw new IllegalArgumentException("In valid group name. The group name must meet the regex: " + this.patternService.getGroupNameRegex());
        }
        GroupMetaEntity entity = new GroupMetaEntity(groupName, owner);

        if (this.groupMetaRepository.existsByGroupName(groupName)) {
            throw new DuplicatedEntityException(groupName);
        }
        this.groupMetaRepository.save(entity);
    }

    @Transactional
    public void deleteGroup(String groupName) {
        if (!this.groupMetaRepository.existsByGroupName(groupName)) {
            throw new EntityNotFoundException(groupName);
        }
        // delete all member
        this.groupUserRepository.deleteAllByGroupName(groupName);
        // delete meta data
        this.groupMetaRepository.deleteByGroupName(groupName);
    }

    public GroupDetailModel queryGroupMeta(String groupName) {
        GroupMetaEntity entity = this.groupMetaRepository.findByGroupName(groupName);
        if (entity == null) {
            return null;
        }

        return new GroupDetailModel(
                entity.getGroupName(),
                entity.getOwner(),
                GeneralKt.getUnixTimestamp(entity.getCreatedTime().getTime())
        );
    }

    @Transactional
    public void setGroupOwner(String groupName, String owner) {
        if (!this.groupMetaRepository.existsByGroupName(groupName)) {
            throw new EntityNotFoundException(groupName);
        }
        this.groupMetaRepository.updateGroupOwner(groupName, owner);
    }

    @Transactional
    public void addUserToGroup(String groupName, String username) {
        if (!this.groupMetaRepository.existsByGroupName(groupName)) {
            throw new EntityNotFoundException(groupName);
        }
        GroupUserEntity entity = new GroupUserEntity(groupName, username);
        if (this.groupUserRepository.exists(Example.of(entity))) {
            throw new DuplicatedEntityException("group member");
        }
        this.groupUserRepository.save(entity);
    }

    @Transactional
    public void removeUserFromGroup(String groupName, String username) {
        if (!this.groupMetaRepository.existsByGroupName(groupName)) {
            throw new EntityNotFoundException(groupName);
        }
        if (!this.groupUserRepository.existsByGroupNameAndUsername(groupName, username)) {
            throw new EntityNotFoundException("User " + username + " in " + groupName);
        }
        this.groupUserRepository.deleteByGroupNameAndUsername(groupName, username);
    }

    public List<String> listUserOwnedGroup(String username, Pageable pageable) {
        List<String> result = new LinkedList<>();
        this.groupMetaRepository.findAllByOwner(username, pageable)
                .forEach(m -> result.add(m.getGroupName()));
        return result;
    }

    public List<String> listUserJoinedGroup(String username, Pageable pageable) {
        List<String> result = new LinkedList<>();

        this.groupUserRepository.findAllByUsername(username, pageable)
                .forEach(r -> result.add(r.getGroupName()));
        return result;
    }

    public boolean userIsGroupMember(String groupName, String username) {
        return this.groupUserRepository.existsByGroupNameAndUsername(groupName, username);
    }

    public boolean userIsGroupOwner(String groupName, String username) {
        return this.groupMetaRepository.existsByGroupNameAndOwner(groupName, username);
    }

    public List<String> listGroupName(String keyword, Pageable pageable) {
        List<String> result = new LinkedList<>();
        this.groupMetaRepository.findAllByGroupNameContains(keyword, pageable)
                .forEach(m -> result.add(m.getGroupName()));
        return result;
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arstue/GroupService.kt
package info.skyblond.archivedag.arstue

import info.skyblond.archivedag.arstue.entity.GroupMetaEntity
import info.skyblond.archivedag.arstue.entity.GroupUserEntity
import info.skyblond.archivedag.arstue.model.GroupDetailModel
import info.skyblond.archivedag.arstue.repo.GroupMetaRepository
import info.skyblond.archivedag.arstue.repo.GroupUserRepository
import info.skyblond.archivedag.arstue.service.PatternService
import info.skyblond.archivedag.commons.DuplicatedEntityException
import info.skyblond.archivedag.commons.EntityNotFoundException
import info.skyblond.archivedag.commons.getUnixTimestamp
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.util.*
import javax.transaction.Transactional

@Service
class GroupService(
    private val patternService: PatternService,
    private val groupMetaRepository: GroupMetaRepository,
    private val groupUserRepository: GroupUserRepository
) {
    @Transactional
    fun createGroup(groupName: String, owner: String) {
        require(patternService.isValidGroupName(groupName)) { "In valid group name. The group name must meet the regex: " + patternService.groupNameRegex }
        val entity = GroupMetaEntity(groupName, owner)
        if (groupMetaRepository.existsByGroupName(groupName)) {
            throw DuplicatedEntityException(groupName)
        }
        groupMetaRepository.save(entity)
    }

    @Transactional
    fun deleteGroup(groupName: String) {
        if (!groupMetaRepository.existsByGroupName(groupName)) {
            throw EntityNotFoundException(groupName)
        }
        // delete all member
        groupUserRepository.deleteAllByGroupName(groupName)
        // delete meta data
        groupMetaRepository.deleteByGroupName(groupName)
    }

    fun queryGroupMeta(groupName: String): GroupDetailModel {
        val entity = groupMetaRepository.findByGroupName(groupName)
            ?: throw EntityNotFoundException(groupName)
        return GroupDetailModel(
            entity.groupName,
            entity.owner,
            getUnixTimestamp(entity.createdTime.time)
        )
    }

    @Transactional
    fun setGroupOwner(groupName: String, owner: String) {
        if (!groupMetaRepository.existsByGroupName(groupName)) {
            throw EntityNotFoundException(groupName)
        }
        groupMetaRepository.updateGroupOwner(groupName, owner)
    }

    @Transactional
    fun addUserToGroup(groupName: String, username: String) {
        if (!groupMetaRepository.existsByGroupName(groupName)) {
            throw EntityNotFoundException(groupName)
        }
        val entity = GroupUserEntity(groupName, username)
        if (groupUserRepository.existsByGroupNameAndUsername(groupName, username)) {
            throw DuplicatedEntityException("group member")
        }
        groupUserRepository.save(entity)
    }

    @Transactional
    fun removeUserFromGroup(groupName: String, username: String) {
        if (!groupMetaRepository.existsByGroupName(groupName)) {
            throw EntityNotFoundException(groupName)
        }
        if (!groupUserRepository.existsByGroupNameAndUsername(groupName, username)) {
            throw EntityNotFoundException("User $username in $groupName")
        }
        groupUserRepository.deleteByGroupNameAndUsername(groupName, username)
    }

    fun listUserOwnedGroup(username: String, pageable: Pageable): List<String> {
        val result: MutableList<String> = LinkedList()
        groupMetaRepository.findAllByOwner(username, pageable)
            .forEach { result.add(it.groupName) }
        return result
    }

    fun listUserJoinedGroup(username: String, pageable: Pageable): List<String> {
        val result: MutableList<String> = LinkedList()
        groupUserRepository.findAllByUsername(username, pageable)
            .forEach { result.add(it.groupName) }
        return result
    }

    fun userIsGroupMember(groupName: String, username: String): Boolean {
        return groupUserRepository.existsByGroupNameAndUsername(groupName, username)
    }

    fun userIsGroupOwner(groupName: String, username: String): Boolean {
        return groupMetaRepository.existsByGroupNameAndOwner(groupName, username)
    }

    fun listGroupName(keyword: String, pageable: Pageable): List<String> {
        val result: MutableList<String> = LinkedList()
        groupMetaRepository.findAllByGroupNameContains(keyword, pageable)
            .forEach { result.add(it.groupName) }
        return result
    }
}
